core/deployment/src/main/java/io/quarkiverse/operatorsdk/deployment/AddClusterRolesDecorator.java

```diff
@@ -100,17 +121,59 @@ public static ClusterRole createClusterRole(QuarkusControllerConfiguration<?> cr
                         dependentRule.addToVerbs(RBACVerbs.PATCH);
                     }
                 }
-                clusterRoleBuilder.addToRules(dependentRule.build());
+                collectedRules.add(dependentRule.build());
             }
 
         });
-
         // add additional RBAC rules
-        clusterRoleBuilder.addAllToRules(cri.getAdditionalRBACRules());
+        collectedRules.addAll(cri.getAdditionalRBACRules());
+        Set<PolicyRule> normalizedRules = getNormalizedRules(collectedRules);
+        clusterRoleBuilder.addToRules(normalizedRules.toArray(PolicyRule[]::new));
 
         return clusterRoleBuilder.build();
     }
 
+    @NotNull
+    private static Set<PolicyRule> getNormalizedRules(Set<PolicyRule> collectedRules) {
```

Comment:
Can you add tests for this method and maybe add more comments about what it's doing?